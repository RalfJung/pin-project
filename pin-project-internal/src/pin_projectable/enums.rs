use proc_macro2::{Ident, Span, TokenStream};
use quote::{quote, ToTokens};
use syn::{Field, Fields, FieldsNamed, FieldsUnnamed, ItemEnum, Result, Variant};

use crate::utils::{proj_ident, VecExt};

use super::*;

pub(super) fn parse(
    args: TokenStream,
    mut item: ItemEnum,
    pinned_drop: Option<ItemFn>,
) -> Result<TokenStream> {
    let impl_drop = ImplDrop::new(item.generics.clone(), pinned_drop)?;
    let mut impl_unpin = ImplUnpin::new(args, &item.generics)?;

    if item.variants.is_empty() {
        return Err(error!(item, "cannot be implemented for enums without variants"));
    } else if let Some(e) = item.variants.iter().find_map(|v| {
        v.discriminant
            .as_ref()
            .map(|(_, e)| error!(e, "cannot be implemented for enums with discriminants"))
    }) {
        return Err(e);
    }

    let proj_ident = proj_ident(&item.ident);
    let (proj_item_body, proj_arms) = variants(&mut item, &proj_ident, &mut impl_unpin);

    let ident = &item.ident;
    let proj_generics = proj_generics(&item.generics);
    let proj_ty_generics = proj_generics.split_for_impl().1;
    let (impl_generics, ty_generics, where_clause) = item.generics.split_for_impl();

    let mut proj_items = quote! {
        enum #proj_ident #proj_generics #where_clause #proj_item_body
    };

    proj_items.extend(impl_drop.build(ident));
    proj_items.extend(impl_unpin.build(ident));
    proj_items.extend(quote! {
        impl #impl_generics #ident #ty_generics #where_clause {
            fn project<'__a>(self: ::core::pin::Pin<&'__a mut Self>) -> #proj_ident #proj_ty_generics {
                unsafe {
                    match ::core::pin::Pin::get_unchecked_mut(self) {
                        #proj_arms
                    }
                }
            }
        }
    });

    let mut item = item.into_token_stream();
    item.extend(proj_items);
    Ok(item)
}

fn variants(
    ItemEnum { variants, ident: enum_ident, .. }: &mut ItemEnum,
    proj_ident: &Ident,
    impl_unpin: &mut ImplUnpin,
) -> (TokenStream, TokenStream) {
    let mut arm_vec = Vec::with_capacity(variants.len());
    let mut ty_vec = Vec::with_capacity(variants.len());
    variants.iter_mut().for_each(|Variant { fields, ident, .. }| {
        let (proj_arm, proj_ty) = match fields {
            Fields::Unnamed(fields) => unnamed(fields, ident, enum_ident, proj_ident, impl_unpin),
            Fields::Named(fields) => named(fields, ident, enum_ident, proj_ident, impl_unpin),
            Fields::Unit => unit(ident, enum_ident, proj_ident),
        };

        arm_vec.push(proj_arm);
        ty_vec.push(proj_ty);
    });

    let proj_item_body = quote!({ #(#ty_vec,)* });
    let proj_arms = quote!(#(#arm_vec,)*);

    (proj_item_body, proj_arms)
}

fn named(
    FieldsNamed { named: fields, .. }: &mut FieldsNamed,
    variant_ident: &Ident,
    enum_ident: &Ident,
    proj_ident: &Ident,
    impl_unpin: &mut ImplUnpin,
) -> (TokenStream, TokenStream) {
    let mut pat_vec = Vec::with_capacity(fields.len());
    let mut expr_vec = Vec::with_capacity(fields.len());
    let mut ty_vec = Vec::with_capacity(fields.len());
    fields.iter_mut().for_each(|Field { attrs, ident, ty, .. }| {
        if attrs.find_remove(PIN) {
            impl_unpin.push(ty);
            expr_vec.push(quote!(#ident: ::core::pin::Pin::new_unchecked(#ident)));
            ty_vec.push(quote!(#ident: ::core::pin::Pin<&'__a mut #ty>));
        } else {
            expr_vec.push(quote!(#ident: #ident));
            ty_vec.push(quote!(#ident: &'__a mut #ty));
        }

        pat_vec.push(ident);
    });

    let proj_arm = quote! {
        #enum_ident::#variant_ident { #(#pat_vec),* } => #proj_ident::#variant_ident { #(#expr_vec),* }
    };
    let proj_ty = quote!(#variant_ident { #(#ty_vec),* });

    (proj_arm, proj_ty)
}

fn unnamed(
    FieldsUnnamed { unnamed: fields, .. }: &mut FieldsUnnamed,
    variant_ident: &Ident,
    enum_ident: &Ident,
    proj_ident: &Ident,
    impl_unpin: &mut ImplUnpin,
) -> (TokenStream, TokenStream) {
    let mut pat_vec = Vec::with_capacity(fields.len());
    let mut expr_vec = Vec::with_capacity(fields.len());
    let mut ty_vec = Vec::with_capacity(fields.len());
    fields.iter_mut().enumerate().for_each(|(i, Field { attrs, ty, .. })| {
        let x = Ident::new(&format!("_x{}", i), Span::call_site());

        if attrs.find_remove(PIN) {
            impl_unpin.push(ty);
            expr_vec.push(quote!(::core::pin::Pin::new_unchecked(#x)));
            ty_vec.push(quote!(::core::pin::Pin<&'__a mut #ty>));
        } else {
            expr_vec.push(quote!(#x));
            ty_vec.push(quote!(&'__a mut #ty));
        }

        pat_vec.push(x);
    });

    let proj_arm = quote! {
        #enum_ident::#variant_ident(#(#pat_vec),*) => #proj_ident::#variant_ident(#(#expr_vec),*)
    };
    let proj_ty = quote!(#variant_ident(#(#ty_vec),*));

    (proj_arm, proj_ty)
}

fn unit(
    variant_ident: &Ident,
    enum_ident: &Ident,
    proj_ident: &Ident,
) -> (TokenStream, TokenStream) {
    let proj_arm = quote! {
        #enum_ident::#variant_ident => #proj_ident::#variant_ident
    };
    let proj_ty = quote!(#variant_ident);

    (proj_arm, proj_ty)
}
