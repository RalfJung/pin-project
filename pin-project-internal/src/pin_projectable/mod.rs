use proc_macro2::{Ident, Span, TokenStream};
use quote::{quote, ToTokens};
use syn::parse::{Parse, ParseStream};
use syn::{
    Attribute, Generics, Item, ItemEnum, ItemFn, ItemStruct, Meta, NestedMeta, Result, Type,
};

use crate::utils::VecExt;

mod enums;
mod structs;

/// The annotation for pinned type.
const PIN: &str = "pin";

const PINNED_DROP: &str = "pinned_drop";

struct PinProject {
    items: Vec<Item>,
}

impl Parse for PinProject {
    fn parse(input: ParseStream<'_>) -> Result<Self> {
        let mut items = vec![];
        while !input.is_empty() {
            items.push(input.parse()?);
        }
        Ok(PinProject { items })
    }
}

fn handle_type(args: TokenStream, item: Item, pinned_drop: Option<ItemFn>) -> Result<TokenStream> {
    match item {
        Item::Struct(item) => {
            ensure_not_packed(&item.attrs)?;
            Ok(structs::parse(args, item, pinned_drop)?)
        }
        Item::Enum(item) => {
            ensure_not_packed(&item.attrs)?;
            Ok(enums::parse(args, item, pinned_drop)?)
        }
        _ => panic!("Unexpected item: {:?}", item),
    }
}

pub(super) fn pin_project(input: TokenStream) -> Result<TokenStream> {
    let span = span!(input);
    let items: Vec<Item> = syn::parse2::<PinProject>(input)?.items;

    let mut found_type = None;
    let mut found_pinned_drop = None;

    for mut item in items {
        match &mut item {
            Item::Struct(ItemStruct { attrs, .. }) | Item::Enum(ItemEnum { attrs, .. }) => {
                if found_type.is_none() {
                    if let Some(pos) = attrs.iter().position(|a| a.path.is_ident("pin_projectable")) {
                        // Remove the 'pin_projectable' attribute, to prevent it from
                        // being parsed again
                        let attr = attrs.remove(pos);
                        let args = match attr.parse_meta()? {
                            Meta::List(l) => l.nested.into_token_stream(),
                            Meta::Word(_) => TokenStream::new(),
                            _ => return Err(error!(span!(attr), "invalid arguments"))
                        };

                        found_type = Some((item.clone(), args));
                    } else {
                        return Err(error!(span, "type delcared in pin_project! must have pin_projectable attribute"))
                    }
                } else {
                    return Err(error!(span, "cannot declare multiple types within pinned module"))
                }
            },
            Item::Fn(ref mut fn_) => {
                if fn_.attrs.find_remove(PINNED_DROP) {
                    if found_pinned_drop.is_none() {
                        found_pinned_drop = Some(fn_.clone());
                    } else {
                        return Err(error!(span, "cannot declare multiple functions within pinned module"));
                    }
                } else {
                    return Err(error!(span, "only #[pinned_drop] functions cannot be declared within pinend module"));
                }
            },
            _ => return Err(error!(span, "pinned module may only contain a struct/enum with an option #[pinned_drop] function"))
        }
    }

    if found_type.is_none() {
        return Err(error!(span, "pin_project must declare a struct or enum"));
    }

    let (type_, args) = match found_type {
        Some(t) => t,
        None => return Err(error!(span, "No #[pin_projectable] type found!")),
    };

    let res = handle_type(args, type_, found_pinned_drop.clone());
    res
}

pub(super) fn attribute(args: TokenStream, input: TokenStream) -> Result<TokenStream> {
    let span = span!(input);
    let item = syn::parse2(input)?;
    match &item {
        Item::Struct(_) | Item::Enum(_) => handle_type(args, item, None),
        _ => Err(error!(span, "may only be used on structs or enums")),
    }
}

fn ensure_not_packed(attrs: &[Attribute]) -> Result<()> {
    for attr in attrs {
        if let Ok(meta) = attr.parse_meta() {
            if let Meta::List(l) = meta {
                if l.ident == "repr" {
                    for repr in l.nested.iter() {
                        if let NestedMeta::Meta(Meta::Word(w)) = repr {
                            if w == "packed" {
                                return Err(error!(
                                    w,
                                    "pin_projectable may not be used on #[repr(packed)] types"
                                ));
                            }
                        }
                    }
                }
            }
        }
    }
    return Ok(());
}

/// Makes the generics of projected type from the reference of the original generics.
fn proj_generics(generics: &Generics) -> Generics {
    let mut generics = generics.clone();
    generics.params.insert(0, syn::parse_quote!('__a));
    generics
}

struct ImplDrop {
    generics: Generics,
    pinned_drop: Option<ItemFn>,
}

impl ImplDrop {
    /// Parses attribute arguments.
    fn new(generics: Generics, pinned_drop: Option<ItemFn>) -> Result<Self> {
        Ok(ImplDrop { generics, pinned_drop })
    }

    fn build(self, ident: &Ident) -> TokenStream {
        let (impl_generics, ty_generics, where_clause) = self.generics.split_for_impl();

        match self.pinned_drop {
            Some(fn_) => {
                let fn_name = fn_.ident.clone();
                quote! {
                    impl #impl_generics ::core::ops::Drop for #ident #ty_generics #where_clause {
                        fn drop(&mut self) {
                            // Declare the #[pinned_drop] function *inside* our drop function
                            // This guarantees that it's impossible for any other user code
                            // to call it
                            #fn_
                            // Safety - we're in 'drop', so we know that 'self' will
                            // never move again
                            let pinned_self = unsafe { ::core::pin::Pin::new_unchecked(self) };
                            // 'pinned_drop' is a free function - if it were part of a trait impl,
                            // it would be possible for user code to call it by directly invoking
                            // the trait.
                            // Therefore, we enforce a return type of '()' by explicitly
                            // assigning it to a temporary.
                            let _: () = #fn_name(pinned_self);
                        }
                    }
                }
            }
            None => {
                quote! {
                    impl #impl_generics ::core::ops::Drop for #ident #ty_generics #where_clause {
                        fn drop(&mut self) {
                            // Do nothing. The precense of this Drop
                            // impl ensures that the user can't provide one of their own
                        }
                    }
                }
            }
        }
    }
}

// =================================================================================================
// conditional Unpin implementation

struct ImplUnpin {
    generics: Generics,
    auto: bool,
}

impl ImplUnpin {
    /// Parses attribute arguments.
    fn new(args: TokenStream, generics: &Generics) -> Result<Self> {
        let mut generics = generics.clone();
        generics.make_where_clause();

        match &*args.to_string() {
            "" => Ok(Self { generics: generics.clone(), auto: true }),
            "unsafe_Unpin" => Ok(Self { generics: generics.clone(), auto: false }),
            _ => Err(error!(args, "an invalid argument was passed")),
        }
    }

    fn push(&mut self, ty: &Type) {
        // We only add bounds for automatically generated impls
        if self.auto {
            self.generics
                .make_where_clause()
                .predicates
                .push(syn::parse_quote!(#ty: ::core::marker::Unpin));
        }
    }

    /// Creates `Unpin` implementation.
    fn build(self, ident: &Ident) -> TokenStream {
        let (impl_generics, ty_generics, where_clause) = self.generics.split_for_impl();
        let mut where_clause = where_clause.unwrap().clone(); // Created in 'new'

        let res = if self.auto {
            quote! {
                impl #impl_generics ::core::marker::Unpin for #ident #ty_generics #where_clause {}
            }
        } else {
            let pin_project_crate = pin_project_crate_path();
            where_clause.predicates.push(syn::parse_quote!(::#pin_project_crate::Wrapper<#ident #ty_generics>: ::#pin_project_crate::UnsafeUnpin));

            quote! {
                impl #impl_generics ::core::marker::Unpin for #ident #ty_generics #where_clause {}
            }
        };
        res
    }
}

/// If the 'renamed' feature is enabled, we detect
/// the actual name of the 'pin-project' crate in the consumer's Cargo.toml
#[cfg(feature = "renamed")]
fn pin_project_crate_path() -> Ident {
    use crate::PIN_PROJECT_CRATE;
    // This is fairly subtle.
    // Normally, you would use `env!("CARGO_PKG_NAME")` to get the name of the package,
    // since it's set at compile time.
    // However, we're in a proc macro, which runs while *another* crate is being compiled.
    // By retreiving the runtime value of `CARGO_PKG_NAME`, we can figure out the name
    // of the crate that's calling us.
    let cur_crate = std::env::var("CARGO_PKG_NAME")
        .expect("Could not find CARGO_PKG_NAME environemnt variable");
    Ident::new(
        if cur_crate == "pin-project" { "pin_project" } else { PIN_PROJECT_CRATE.as_str() },
        Span::call_site(),
    )
}

/// If the 'renamed' feature is not enabled, we just
/// assume that the 'pin-project' dependency has not been renamed
#[cfg(not(feature = "renamed"))]
fn pin_project_crate_path() -> Ident {
    Ident::new("pin_project", Span::call_site())
}
