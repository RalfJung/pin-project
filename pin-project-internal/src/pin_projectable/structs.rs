use proc_macro2::TokenStream;
use quote::{quote, ToTokens};
use syn::{Field, Fields, FieldsNamed, FieldsUnnamed, Index, ItemStruct, Result};

use crate::utils::{proj_ident, VecExt};

use super::*;

pub(super) fn parse(
    args: TokenStream,
    mut item: ItemStruct,
    pinned_drop: Option<ItemFn>,
) -> Result<TokenStream> {
    let impl_drop = ImplDrop::new(item.generics.clone(), pinned_drop)?;
    let mut impl_unpin = ImplUnpin::new(args, &item.generics)?;

    let (proj_item_body, proj_init_body) = match &mut item.fields {
        Fields::Named(FieldsNamed { named: fields, .. })
        | Fields::Unnamed(FieldsUnnamed { unnamed: fields, .. })
            if fields.is_empty() =>
        {
            return Err(error!(item.fields, "cannot be implemented for structs with zero fields"))
        }
        Fields::Unit => return Err(error!(item, "cannot be implemented for structs with units")),

        Fields::Named(fields) => named(fields, &mut impl_unpin),
        Fields::Unnamed(fields) => unnamed(fields, &mut impl_unpin),
    };

    let ident = &item.ident;
    let proj_ident = proj_ident(&item.ident);
    let proj_generics = proj_generics(&item.generics);
    let proj_ty_generics = proj_generics.split_for_impl().1;
    let (impl_generics, ty_generics, where_clause) = item.generics.split_for_impl();

    let mut proj_items = quote! {
        struct #proj_ident #proj_generics #where_clause #proj_item_body
    };

    proj_items.extend(impl_drop.build(ident));
    proj_items.extend(impl_unpin.build(ident));
    proj_items.extend(quote! {
        impl #impl_generics #ident #ty_generics #where_clause {
            fn project<'__a>(self: ::core::pin::Pin<&'__a mut Self>) -> #proj_ident #proj_ty_generics {
                unsafe {
                    let this = ::core::pin::Pin::get_unchecked_mut(self);
                    #proj_ident #proj_init_body
                }
            }
        }
    });

    let mut item = item.into_token_stream();
    item.extend(proj_items);
    Ok(item)
}

fn named(
    FieldsNamed { named: fields, .. }: &mut FieldsNamed,
    impl_unpin: &mut ImplUnpin,
) -> (TokenStream, TokenStream) {
    let mut proj_fields = Vec::with_capacity(fields.len());
    let mut proj_init = Vec::with_capacity(fields.len());
    fields.iter_mut().for_each(|Field { attrs, ident, ty, .. }| {
        if attrs.find_remove(PIN) {
            impl_unpin.push(ty);
            proj_fields.push(quote!(#ident: ::core::pin::Pin<&'__a mut #ty>));
            proj_init.push(quote!(#ident: ::core::pin::Pin::new_unchecked(&mut this.#ident)));
        } else {
            proj_fields.push(quote!(#ident: &'__a mut #ty));
            proj_init.push(quote!(#ident: &mut this.#ident));
        }
    });

    let proj_item_body = quote!({ #(#proj_fields,)* });
    let proj_init_body = quote!({ #(#proj_init,)* });

    (proj_item_body, proj_init_body)
}

fn unnamed(
    FieldsUnnamed { unnamed: fields, .. }: &mut FieldsUnnamed,
    impl_unpin: &mut ImplUnpin,
) -> (TokenStream, TokenStream) {
    let mut proj_fields = Vec::with_capacity(fields.len());
    let mut proj_init = Vec::with_capacity(fields.len());
    fields.iter_mut().enumerate().for_each(|(i, Field { attrs, ty, .. })| {
        let i = Index::from(i);
        if attrs.find_remove(PIN) {
            impl_unpin.push(ty);
            proj_fields.push(quote!(::core::pin::Pin<&'__a mut #ty>));
            proj_init.push(quote!(::core::pin::Pin::new_unchecked(&mut this.#i)));
        } else {
            proj_fields.push(quote!(&'__a mut #ty));
            proj_init.push(quote!(&mut this.#i));
        }
    });

    let proj_item_body = quote!((#(#proj_fields,)*););
    let proj_init_body = quote!((#(#proj_init,)*));

    (proj_item_body, proj_init_body)
}
