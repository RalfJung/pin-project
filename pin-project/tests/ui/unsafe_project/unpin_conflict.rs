// compile-fail

#![deny(warnings)]

use pin_project::pin_projectable;
use std::pin::Pin;

// The same implementation.

#[pin_projectable] //~ ERROR E0119
struct Foo<T, U> {
    #[pin]
    future: T,
    field: U,
}

impl<T, U> Foo<T, U> {
    fn baz(mut self: Pin<&mut Self>) {
        let this = self.project();
        let _: Pin<&mut T> = this.future; // Pinned reference to the field
        let _: &mut U = this.field; // Normal reference to the field
    }
}

// conflicting implementations
impl<T, U> Unpin for Foo<T, U> where T: Unpin {} // Conditional Unpin impl

// The implementation that under different conditions.

#[pin_projectable] //~ ERROR E0119
struct Bar<T, U> {
    #[pin]
    future: T,
    field: U,
}

impl<T, U> Bar<T, U> {
    fn baz(mut self: Pin<&mut Self>) {
        let this = self.project();
        let _: Pin<&mut T> = this.future; // Pinned reference to the field
        let _: &mut U = this.field; // Normal reference to the field
    }
}

// conflicting implementations
impl<T, U> Unpin for Bar<T, U> {} // Non-conditional Unpin impl

#[pin_projectable] //~ ERROR E0119
struct Baz<T, U> {
    #[pin]
    future: T,
    field: U,
}

impl<T, U> Baz<T, U> {
    fn baz(mut self: Pin<&mut Self>) {
        let this = self.project();
        let _: Pin<&mut T> = this.future; // Pinned reference to the field
        let _: &mut U = this.field; // Normal reference to the field
    }
}

// conflicting implementations
impl<T: Unpin, U: Unpin> Unpin for Baz<T, U> {} // Conditional Unpin impl

fn main() {}
