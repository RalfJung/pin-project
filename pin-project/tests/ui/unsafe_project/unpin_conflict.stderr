error[E0119]: conflicting implementations of trait `std::marker::Unpin` for type `Foo<_, _>`:
  --> $DIR/unpin_conflict.rs:10:1
   |
10 | #[pin_projectable] //~ ERROR E0119
   | ^^^^^^^^^^^^^^^^^^ conflicting implementation for `Foo<_, _>`
...
26 | impl<T, U> Unpin for Foo<T, U> where T: Unpin {} // Conditional Unpin impl
   | --------------------------------------------- first implementation here

error[E0119]: conflicting implementations of trait `std::marker::Unpin` for type `Bar<_, _>`:
  --> $DIR/unpin_conflict.rs:30:1
   |
30 | #[pin_projectable] //~ ERROR E0119
   | ^^^^^^^^^^^^^^^^^^ conflicting implementation for `Bar<_, _>`
...
46 | impl<T, U> Unpin for Bar<T, U> {} // Non-conditional Unpin impl
   | ------------------------------ first implementation here

error[E0119]: conflicting implementations of trait `std::marker::Unpin` for type `Baz<_, _>`:
  --> $DIR/unpin_conflict.rs:48:1
   |
48 | #[pin_projectable] //~ ERROR E0119
   | ^^^^^^^^^^^^^^^^^^ conflicting implementation for `Baz<_, _>`
...
64 | impl<T: Unpin, U: Unpin> Unpin for Baz<T, U> {} // Conditional Unpin impl
   | -------------------------------------------- first implementation here

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0119`.
